src/GSoft.Extensions.Mongo.ApplicationInsights/ApplicationInsightsEventSubscriber.cs

```diff
@@ -0,0 +1,159 @@
+        {
+            IPEndPoint endPoint => endPoint.Address + ":" + endPoint.Port,
+            DnsEndPoint endPoint => endPoint.Host + ":" + endPoint.Port,
+            _ => "unknown",
+        };
+
+        if (this._options.Telemetry.CaptureCommandText)
+        {
+            operation.Telemetry.Data = evt.Command.ToString();
+        }
+
+        // Originating activity must be captured AFTER that the operation is created
+        // Because ApplicationInsights SDK creates another intermediate Activity
+        var originatingActivity = Activity.Current;
+
+        this._currentOperationsMap.TryAdd(evt.RequestId, new ActivityAwareOperationHolder(operation, originatingActivity));
+    }
+
+    private void CommandSucceededEventHandler(CommandSucceededEvent evt)
+    {
+        if (!this._currentOperationsMap.TryRemove(evt.RequestId, out var operation))
+        {
+            return;
+        }
+
+        using (operation)
+        {
+            operation.Telemetry.Duration = evt.Duration;
+            operation.Telemetry.Success = true;
+        }
+    }
+
+    private void CommandFailedEventHandler(CommandFailedEvent evt)
+    {
+        if (!this._currentOperationsMap.TryRemove(evt.RequestId, out var operation))
+        {
+            return;
+        }
+
+        using (operation)
+        {
+            operation.Telemetry.Duration = evt.Duration;
+            operation.Telemetry.Success = false;
+
+            if (!operation.Telemetry.Properties.ContainsKey("Exception"))
+            {
+                operation.Telemetry.Properties.Add("Exception", evt.Failure.ToString());
+            }
+        }
+    }
+
+    private IOperationHolder<DependencyTelemetry> StartActivityAwareDependencyOperation(string name)
+    {
+        if (Activity.Current is { } activity && TracingHelper.IsMongoActivity(activity))
+        {
+            // When the current activity is our own Mongo activity created in our previous command tracing behavior,
+            // then we use it to initialize the Application Insights operation.
+            // The Application Insights SDK will take care of populating the parent-child relationship
+            // and bridge the gap between our activity, its own internal activity and the AI operation telemetry.
+            // Not doing that could cause some Application Insights AND OpenTelemetry spans to be orphans.
+            var operation = this._telemetryClient.StartOperation<DependencyTelemetry>(activity);
+            operation.Telemetry.Name = name;
+
+            // Remove telemetry copied from our Mongo activity as we already have it in the AI operation
+            foreach (var item in activity.Baggage)
+            {
+                operation.Telemetry.Properties.Remove(item.Key);
+            }
+
+            foreach (var item in activity.Tags)
+            {
+                operation.Telemetry.Properties.Remove(item.Key);
+            }
```

Comment:
Given we're removing a bunch of keys, could we do something like such to iterate only once (I didn't check whether types are compatible):
```suggestion
            foreach (var item in activity.Baggage.Union(activity.Tags))
            {
                operation.Telemetry.Properties.Remove(item.Key);
            }
```